File: de.fhdo.lemma.live_validation/client/src/main/java/de/fhdo/lemma/live_validation/client/AbstractLiveValidator.java
package de.fhdo.lemma.live_validation.client;

import de.fhdo.lemma.live_validation.protocol.extensions.PublishIssuesParams;

import java.util.Set;
import java.util.concurrent.CompletableFuture;

/**
 * Abstract superclass for Live Validators, i.e., components that perform a Live Validation.
 *
 * @author <a href="mailto:florian.rademacher@fh-dortmund.de">Florian Rademacher</a>
 */
public abstract class AbstractLiveValidator extends BaseClient {
    /**
     * Retrieve supported file extensions to identify the file formats a concrete Live Validator may handle
     */
    public abstract Set<String> getSupportedFileExtensions();

    /**
     * Convenience implementation of LiveValidationLanguageClient.requestValidatorRegistration. Internally invokes
     * getSupportedFileExtensions.
     */
    @Override
    public final CompletableFuture<Set<String>> requestValidatorRegistration() {
        return CompletableFuture.completedFuture(getSupportedFileExtensions());
    }

    /**
     * NOOP implementation of LiveValidationLanguageClient.publishIssues, because Live Validators never simultaneously
     * act as Live Validation initiators
     */
    @Override
    public final void publishIssues(PublishIssuesParams publishIssuesParams) {
        // NOOP
    }
}

File: de.fhdo.ddmm.live_validation/src/main/kotlin/de/fhdo/ddmm/live_validation/clients/AbstractLiveValidator.kt
package de.fhdo.ddmm.live_validation.clients

import de.fhdo.ddmm.live_validation.protocol.LiveValidationLanguageClient
import de.fhdo.ddmm.live_validation.protocol.PublishIssuesParams
import java.io.OutputStream
import java.util.concurrent.CompletableFuture

/**
 * Abstract superclass for Live Validators, i.e., components that perform a Live Validation.
 *
 * @author [Florian Rademacher](mailto:florian.rademacher@fh-dortmund.de)
 */
abstract class AbstractLiveValidator : BaseClient(), LiveValidationLanguageClient {
    /**
     * Establish connection to a Live Validation server
     */
    @JvmOverloads fun start(hostname: String, port: Int, rootUri: String? = null, traceStream: OutputStream? = null) {
        super.start(hostname, port, this, rootUri, traceStream)
    }

    /**
     * Retrieve supported file extensions to identify the file formats a concrete Live Validator may handle
     */
    abstract fun getSupportedFileExtensions() : Set<String>

    /**
     * Convenience implementation of [LiveValidationLanguageClient.requestValidatorRegistration]. Internally invokes
     * [getSupportedFileExtensions].
     */
    final override fun requestValidatorRegistration(): CompletableFuture<Set<String>>
        = CompletableFuture.completedFuture(getSupportedFileExtensions())

    /**
     * NOOP implementation of [LiveValidationLanguageClient.publishIssues], because Live Validators never simultaneously
     * act as Live Validation initiators, i.e., hence validation result receivers
     */
    final override fun publishIssues(publishIssuesParams: PublishIssuesParams) { }
}